src/Runner.Worker/Container/ContainerHooks/ContainerHookManager.cs

```diff
@@ -0,0 +1,308 @@
+        Task RunContainerStepAsync(IExecutionContext context, ContainerInfo container, string dockerFile);
+        Task RunScriptStepAsync(IExecutionContext context, ContainerInfo container, string arguments, string fileName, IDictionary<string, string> environment, string prependPath, string workingDirectory);
+        string GetContainerHookData();
+    }
+
+    public class ContainerHookManager : RunnerService, IContainerHookManager
+    {
+        private const string ResponseFolderName = "_runner_hook_responses";
+        private string HookIndexPath;
+
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+            HookIndexPath = $"{Environment.GetEnvironmentVariable(Constants.Hooks.ContainerHooksPath)}";
+            ValidateHookExecutable();
+        }
+
+        public async Task PrepareJobAsync(IExecutionContext context, List<ContainerInfo> containers)
+        {
+            Trace.Entering();
+            var jobContainer = containers.Where(c => c.IsJobContainer).FirstOrDefault();
+            var serviceContainers = containers.Where(c => c.IsJobContainer == false).ToList();
+
+            var input = new HookInput
+            {
+                Command = HookCommand.PrepareJob,
+                ResponseFile = GenerateResponsePath(),
+                Args = new PrepareJobArgs
+                {
+                    Container = jobContainer.GetHookContainer(),
+                    Services = serviceContainers.Select(c => c.GetHookContainer()).ToList(),
+                }
+            };
+
+            var prependPath = GetPrependPath(context);
+            PrepareJobResponse response;
+            try
+            {
+                response = await ExecuteHookScript<PrepareJobResponse>(context, input, ActionRunStage.Pre, prependPath);
+            }
+            catch (Exception ex)
+            {
+                Trace.Error(ex);
+                throw new Exception($"Custom container implementation failed with error: {ex.Message}. Please contact your self hosted runner administrator.");
+            }
+            jobContainer.IsAlpine = response.IsAlpine.Value;
+            SaveHookState(context, response.State, input);
+            UpdateJobContext(context, jobContainer, serviceContainers, response);
+        }
+
+        public async Task CleanupJobAsync(IExecutionContext context, List<ContainerInfo> containers)
+        {
+            Trace.Entering();
+            var input = new HookInput
+            {
+                Command = HookCommand.CleanupJob,
+                ResponseFile = GenerateResponsePath(),
+                State = context.Global.ContainerHookState,
+            };
+            var prependPath = GetPrependPath(context);
+            try
+            {
+                await ExecuteHookScript<PrepareJobResponse>(context, input, ActionRunStage.Pre, prependPath);
```

Comment:
also, we didn't use the return value, is it expected?

Reply:
Should be HookResponse, fixed it.

Aside from PrepareJobResponse, we only use the response to update `State`. After cleanupJob, we don't need to update it anymore because the job is finished.